db/addons.go

```diff
@@ -163,3 +197,195 @@ func (d *Database) DeleteAddon(ctx context.Context, addonID string, opts ...Quer
+	_, db := d.querySettings(opts...)
+
+	ds := subAddonDS(db).
+		Where(t.SubscriptionAddons.Col("id").Eq(subAddonID)).
+		Executor()
+	d.LogSQL(ds)
+
+	subAddon := &SubscriptionAddon{}
+	if _, err := ds.ScanStructContext(ctx, subAddon); err != nil {
+		return nil, err
+	}
+
+	return subAddon, nil
+}
+
+func (d *Database) ListSubscriptionAddons(ctx context.Context, subscriptionID string, opts ...QueryOption) ([]SubscriptionAddon, error) {
+	_, db := d.querySettings(opts...)
+
+	ds := subAddonDS(db).Executor()
+	d.LogSQL(ds)
+
+	var addons []SubscriptionAddon
+	if err := ds.ScanStructsContext(ctx, &addons); err != nil {
+		return nil, errors.Wrap(err, "unable to list addons")
+	}
+
+	return addons, nil
+}
+
+func (d *Database) AddSubscriptionAddon(ctx context.Context, subscriptionID, addonID string, opts ...QueryOption) (*SubscriptionAddon, error) {
+	qs, db, err := d.querySettingsWithTX(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	if qs.doRollback {
+		defer db.Rollback()
+	}
+
+	addon, err := d.GetAddonByID(ctx, addonID, WithTXRollbackCommit(db, false, false))
+	if err != nil {
+		return nil, err
+	}
+
+	ds := db.Insert(t.SubscriptionAddons).
+		Rows(goqu.Record{
+			"subscription_id": subscriptionID,
+			"addon_id":        addonID,
+			"amount":          addon.DefaultAmount,
+			"paid":            addon.DefaultPaid,
+		}).
+		Returning(t.SubscriptionAddons.Col("id")).
+		Executor()
+
+	var newAddonID string
+	if _, err = ds.ScanValContext(ctx, &newAddonID); err != nil {
+		return nil, err
+	}
+
+	subscription, err := d.GetSubscriptionByID(ctx, subscriptionID, WithTXRollbackCommit(db, false, false))
+	if err != nil {
+		return nil, err
+	}
+
+	if qs.doCommit {
+		if err = db.Commit(); err != nil {
+			return nil, err
+		}
+	}
+
+	retval := &SubscriptionAddon{
+		ID:           newAddonID,
+		Addon:        *addon,
+		Subscription: *subscription,
+		Amount:       addon.DefaultAmount,
+		Paid:         addon.DefaultPaid,
+	}
+
+	return retval, nil
+}
+
+func (d *Database) DeleteSubscriptionAddon(ctx context.Context, subAddonID string, opts ...QueryOption) error {
+	_, db := d.querySettings(opts...)
+
+	ds := db.From(t.SubscriptionAddons).
+		Delete().
+		Where(t.SubscriptionAddons.Col("id").Eq(subAddonID)).
+		Executor()
+
+	_, err := ds.ExecContext(ctx)
+	return err
+}
+
+func (d *Database) UpdateSubscriptionAddon(ctx context.Context, updated *UpdateSubscriptionAddon, opts ...QueryOption) (*SubscriptionAddon, error) {
+	qs, db, err := d.querySettingsWithTX(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	if qs.doRollback {
+		defer db.Rollback()
+	}
+
+	rec := goqu.Record{}
+	if updated.UpdateAmount {
+		rec["amount"] = updated.Amount
+	}
+	if updated.UpdatePaid {
+		rec["paid"] = updated.Paid
+	}
+
+	ds := db.Update(t.SubscriptionAddons).
+		Set(rec).
+		Where(t.SubscriptionAddons.Col("id").Eq(updated.ID)).
+		Executor()
+
+	if _, err = ds.ExecContext(ctx); err != nil {
+		return nil, err
+	}
```

Comment:
I think that this might cause problems if a non-existent subscription add-on ID is passed in as well. It might be helpful to check the number of rows affected and return early if the number is 0 (or more generally is anything other than 1, although if the number is greater than 1 then that would probably mean that the database schema is broken)